File: 2024/day01/day01p2.py
if __name__ == "__main__":
    list1 = []
    list2map = {}

    nums = open("day01-data.txt", "r")
    for line in nums:
        if line == "\n":
            continue
        list1.append(int(line[:5]))
        secondnum = int(line[8:13])
        if secondnum in list2map:
            list2map[secondnum] += 1
            continue
        list2map[secondnum] = 1

    similarity = 0
    for num in list1:
        similarity += num * list2map.get(num, 0)
    print(similarity)

File: 2024/day01/day01p2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(void) {
  FILE* fptr = fopen("day01-data.txt", "r");

  if (fptr == NULL) {
    fclose(fptr);
    return 1;
  }

  // HACK: Way too much memory is used here. I'm too lazy to implement an actual
  // hashmap. Oh well!
  unsigned short rightlist_occurances[99999] = {0};
  int leftlist[1000];

  char* line = NULL;
  size_t len;
  ssize_t read;

  char num1s[6];
  char num2s[6];

  unsigned short cursor = 0;

  unsigned int similarity_score = 0;

  while ((read = getline(&line, &len, fptr)) != -1) {
    if (*line == '\n')
      break;
    strncpy(num1s, line, 5);
    strncpy(num2s, line + 8, 5);
    num1s[5] = '\0';
    num2s[5] = '\0';
    leftlist[cursor] = atoi(num1s);
    unsigned int tmp = atoi(num2s);
    rightlist_occurances[tmp]++;
    cursor++;
  }

  fclose(fptr);

  for (unsigned short cursor2 = 0; cursor2 < 1000; cursor2++) {
    similarity_score +=
        leftlist[cursor2] * rightlist_occurances[leftlist[cursor2]];
  }

  if (line)
    free(line);

  printf("Similarity: %d", similarity_score);

  return 0;
}
